File: binarySearch.cpp
#include <vector>
class Solution {
public:
  int search(std::vector<int> &nums, int target) {
    if (nums.size() == 0) {
      return -1;
    }

    int lowerBound = 0;
    int upperBound = nums.size() - 1;

    while (lowerBound <= upperBound) {
      int middle = lowerBound + ((upperBound - lowerBound) / 2);
      if (nums[middle] < target) {
        lowerBound = middle + 1;
      } else if (nums[middle] > target) {
        upperBound = middle - 1;
      } else {
        return middle;
      }
    }

    return -1;
  }
};

File: python/binary_search.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l_ptr = 0
        r_ptr = len(nums) - 1

        while l_ptr <= r_ptr:
            mid = (r_ptr + l_ptr) // 2
            if target > nums[mid]:
                l_ptr = mid + 1
            elif target < nums[mid]:
                r_ptr = mid - 1
            else:
                return mid
        return -1
